src/documents/file_handling.py

```diff
@@ -127,123 +113,287 @@ def generate_unique_filename(doc, archive_filename=False):
+        "added_month_name": local_added.strftime("%B"),
+        "added_month_name_short": local_added.strftime("%b"),
+        "added_day": local_added.strftime("%d"),
+    }
+
+
+def get_basic_metadata_context(
+    document: Document,
+    *,
+    no_value_default: str,
+) -> dict[str, str]:
+    """
+    Given a Document, constructs some basic information about it.  If certain values are not set,
+    they will be replaced with the no_value_default.
+
+    Regardless of set or not, the values will be sanitized
+    """
+    return {
+        "title": pathvalidate.sanitize_filename(
+            document.title,
+            replacement_text="-",
+        ),
+        "correspondent": pathvalidate.sanitize_filename(
+            document.correspondent.name,
+            replacement_text="-",
+        )
+        if document.correspondent
+        else no_value_default,
+        "document_type": pathvalidate.sanitize_filename(
+            document.document_type.name,
+            replacement_text="-",
+        )
+        if document.document_type
+        else no_value_default,
+        "asn": str(document.archive_serial_number)
+        if document.archive_serial_number
+        else no_value_default,
+        "owner_username": document.owner.username
+        if document.owner
+        else no_value_default,
+        "original_name": PurePath(document.original_filename).with_suffix("").name
+        if document.original_filename
+        else no_value_default,
+        "doc_pk": f"{document.pk:07}",
+    }
+
+
+def get_tags_context(tags: Iterable[Tag]) -> dict[str, str | list[str]]:
+    """
+    Given an Iterable of tags, constructs some context from them for usage
+    """
+    return {
+        "tag_list": pathvalidate.sanitize_filename(
+            ",".join(
+                sorted(tag.name for tag in tags),
+            ),
+            replacement_text="-",
+        ),
+        # Assumed to be ordered, but a template could loop through to find what they want
+        "tag_name_list": [x.name for x in tags],
+    }
+
+
+def get_custom_fields_context(
+    custom_fields: Iterable[CustomFieldInstance],
+) -> dict[str, dict[str, dict[str, str]]]:
+    """
+    Given an Iterable of CustomFieldInstance, builds a dictionary mapping the field name
+    to its type and value
+    """
+    return {
+        "custom_fields": {
+            pathvalidate.sanitize_filename(
+                field_instance.field.name,
+                replacement_text="-",
+            ): {
+                "type": pathvalidate.sanitize_filename(
+                    field_instance.field.data_type,
+                    replacement_text="-",
+                ),
+                "value": pathvalidate.sanitize_filename(
+                    str(field_instance.value),
+                    replacement_text="-",
+                ),
+            }
+            for field_instance in custom_fields
+        },
+    }
+
+
+def validate_template_and_render(
```

Comment:
You should extract the templating logic into an own file to keep up with the solid principles. Then we would be able to re use it for titles and preview etc. Here you mixed the logic with path specific one into the file handling. That's not ideal software architecture vise ;)